internal/appsec/appsec.go

```diff
@@ -0,0 +1,291 @@
+		AgentURL string
+		// ServiceConfig is the information about the running service we currently protect.
+		Service ServiceConfig
+		// Tags is the list of tags that should be added to security events (eg. pid, os name, etc.).
+		Tags []string
+		// Hostname of the machine we run in.
+		Hostname string
+		// Version of the Go client library
+		Version string
+
+		// MaxBatchLen is the maximum batch length the event batching loop should use. The event batch is sent when
+		// this length is reached. Defaults to 1024.
+		MaxBatchLen int
+		// MaxBatchStaleTime is the maximum amount of time events are kept in the batch. This allows to send the batch
+		// after this amount of time even if the maximum batch length is not reached yet. Defaults to 1 second.
+		MaxBatchStaleTime time.Duration
+	}
+
+	// ServiceConfig is the optional context about the running service.
+	ServiceConfig struct {
+		// Name of the service.
+		Name string
+		// Version of the service.
+		Version string
+		// Environment of the service (eg. dev, staging, prod, etc.)
+		Environment string
+	}
+)
+
+// Default batching configuration values.
+const (
+	defaultMaxBatchLen       = 1024
+	defaultMaxBatchStaleTime = time.Second
+)
+
+// Default timeout of intake requests.
+const defaultIntakeTimeout = 10 * time.Second
+
+// Start AppSec when the environment variable DD_APPSEC_ENABLED is set to true.
+func Start(cfg *Config) {
+	enabled, err := isEnabled()
+	if err != nil {
+		log.Error("appsec: %v", err)
+		return
+	}
+	if !enabled {
+		return
+	}
+
+	appsec, err := new(
+		cfg,
+	)
+	if err != nil {
+		log.Error("appsec: %v", err)
+		log.Error("appsec: disabled due to a startup error")
+		return
+	}
+	appsec.start()
+	setActiveAppSec(appsec)
+}
+
+// Stop AppSec.
+func Stop() {
+	setActiveAppSec(nil)
+}
+
+var (
+	activeAppSec *appsec
+	mu           sync.Mutex
+)
+
+func setActiveAppSec(a *appsec) {
+	mu.Lock()
+	defer mu.Unlock()
+	if activeAppSec != nil {
+		activeAppSec.stop()
+	}
+	activeAppSec = a
+}
+
+func isEnabled() (bool, error) {
+	enabledStr := os.Getenv("DD_APPSEC_ENABLED")
+	if enabledStr == "" {
+		return false, nil
+	}
+	enabled, err := strconv.ParseBool(enabledStr)
+	if err != nil {
+		return false, fmt.Errorf("could not parse DD_APPSEC_ENABLED value `%s` as a boolean value", enabledStr)
+	}
+	return enabled, nil
+}
+
+type appsec struct {
+	client          *intake.Client
+	eventChan       chan *appsectypes.SecurityEvent
+	wg              sync.WaitGroup
+	cfg             *Config
+	unregisterInstr []dyngo.UnregisterFunc
+}
+
+func new(cfg *Config) (*appsec, error) {
+	intakeClient, err := intake.NewClient(cfg.Client, cfg.AgentURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.Hostname == "" {
+		hostname, err := os.Hostname()
+		if err != nil {
+			log.Warn("unable to look up hostname: %v", err)
+		} else {
+			cfg.Hostname = hostname
+		}
+	}
+
+	if cfg.Service.Name == "" {
+		name, err := os.Executable()
```

Comment:
Indeed, let's always use the tracer config. I removed it from appsec - so it now uses what the tracer gives it which can be an empty string.

cc @knusbaum in case you'd be interested in having the basename of the executable name when the service name is not configured by the user.